File: plc4j/integrations/plc4j-apache-karaf/org.apache.plc4x.merlot.modbus.dev/src/main/java/org/apache/plc4x/merlot/modbus/dev/api/ModbusDevice.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.plc4x.merlot.modbus.dev.api;

import io.netty.buffer.ByteBuf;

public interface ModbusDevice {

    /**
     * Enable / Disables device operation
     * @param Enabled
     */
    public void setEnabled(boolean Enabled);
    
    /**
     * Returns the current status of the device.
     * @return true if the device is enabled, false otherwise.
     */
    public boolean getEnabled();    	
	
    /**
     * Set de the device identification. The firts byte from the Modbus ADU
     * @param ui
     */
    public void setUnitIdentifier(byte UID);
    
    /**
     * Get de the device identification.
     * @return the UID of the device.
     */
    public byte getUnitIdentifier();
    
    /**
     * Set de the device description. 
     * @param description Description of the device.
     */
    public void setUnitDescription(String description);
    
    /**
     * Get de the device Description.
     * @return The description of the device.
     */
    public String getUnitDescription();    
    
    /**
     * Get the "discrete inputs" from the device. Complete reference to memory area.
     * 
     */
    public ByteBuf getDiscreteInputs();
    
    /**
     * Set a particular "discrete input" in the device. Only for internal use.
     * @param register
     * @param state
     */
    public void setDiscreteInput(int register,  boolean state);
    
    /**
     * Get a particular "discrete input" in the device. 
     * @param register
     * @param state
     */
    public boolean getDiscreteInput(int register);    
    
    /**
     * Get the "coil" inputs from the device. Complete reference to memory area.
     * 
     */
    public ByteBuf getCoils(); 
    
    /**
     * Set a particular "coil" in the device. Only for internal & external use.
     * @param register
     * @param state
     */
    public void setCoil(int register,  boolean state);
    
    /**
     * Get a particular "discrete input" in the device. 
     * @param register
     */
    public boolean getCoil(int register);  
    
    /**
     * Get the "input registers" inputs from the device. Complete reference to memory area.
     * 
     */
    public ByteBuf getInputRegisters(); 
    
    /**
     * Set a particular "input register" in the device. Only for internal use.
     * @param register
     * @param value
     */
    public void setInputRegister(int register,  short value);
    
    /**
     * Get a particular "input register" in the device. 
     * @param register
     */
    public short getInputRegister(int register);  
    
    /**
     * Get the "input registers" inputs from the device. Complete reference to memory area.
     * 
     */
    public ByteBuf getHoldingRegisters(); 
    
    /**
     * Set a particular "input register" in the device. Only for internal & external use.
     * @param register
     * @param value
     */
    public void setHoldingRegister(int register,  short value);
    
    /**
     * Get a particular "input register" in the device. 
     * @param register
     */
    public short getHoldingRegister(int register); 
    
    /**
     * Perform a dump of the memory contents in hexadecimal / Ascii.
     * @param index Start register address.
     * @param length number of records to be transferred from registers.
     */
    public String HexDump(int index, int length);    
    
    
}

File: sandbox/plc4py/plc4py/drivers/modbus/ModbusDevice.py
#
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
#
import asyncio
import logging
from asyncio import Transport
from dataclasses import dataclass, field
from typing import Dict, List

from bitarray import bitarray

from plc4py.protocols.modbus.readwrite.DataItem import DataItem
from plc4py.protocols.modbus.readwrite.ModbusPDUError import ModbusPDUError
from plc4py.spi.generation.ReadBuffer import ReadBuffer, ReadBufferByteBased

from plc4py.drivers.modbus.ModbusTag import (
    ModbusTagHoldingRegister,
    ModbusTagCoil,
    ModbusTagDiscreteInput,
    ModbusTagInputRegister,
)

from plc4py.api.exceptions.exceptions import PlcRuntimeException
from plc4py.drivers.modbus.ModbusConfiguration import ModbusConfiguration
from plc4py.protocols.modbus.readwrite.ModbusPDUReadCoilsRequest import (
    ModbusPDUReadCoilsRequest,
)
from plc4py.protocols.modbus.readwrite.ModbusPDUReadDiscreteInputsRequest import (
    ModbusPDUReadDiscreteInputsRequest,
)
from plc4py.protocols.modbus.readwrite.ModbusPDUReadInputRegistersRequest import (
    ModbusPDUReadInputRegistersRequest,
)
from plc4py.spi.generation.WriteBuffer import WriteBufferByteBased

from plc4py.api.messages.PlcRequest import PlcReadRequest, PlcWriteRequest
from plc4py.api.messages.PlcResponse import PlcReadResponse, PlcWriteResponse
from plc4py.api.value.PlcValue import PlcValue, PlcResponseCode
from plc4py.protocols.modbus.readwrite.ModbusPDUReadHoldingRegistersRequest import (
    ModbusPDUReadHoldingRegistersRequest,
)
from plc4py.protocols.modbus.readwrite.ModbusTcpADU import ModbusTcpADU
from plc4py.spi.messages.utils.ResponseItem import ResponseItem
from plc4py.spi.values.PlcValues import PlcList, PlcNull
from plc4py.utils.GenericTypes import ByteOrder, AtomicInteger


@dataclass
class ModbusDevice:
    _configuration: ModbusConfiguration
    tags: Dict[str, PlcValue] = field(default_factory=lambda: {})
    _transaction_generator: AtomicInteger = field(
        default_factory=lambda: AtomicInteger()
    )

    async def read(
        self, request: PlcReadRequest, transport: Transport
    ) -> PlcReadResponse:
        """
        Reads one field from the Modbus Device
        """
        if len(request.tags) > 1:
            raise NotImplementedError(
                "The Modbus driver only supports reading single tags at once"
            )
        if len(request.tags) == 0:
            raise PlcRuntimeException("No tags have been specified to read")
        tag = request.tags[request.tag_names[0]]
        logging.debug(f"Reading tag {str(tag)} from Modbus Device")

        # Create future to be returned when a value is returned
        loop = asyncio.get_running_loop()
        message_future = loop.create_future()

        if isinstance(tag, ModbusTagCoil):
            pdu = ModbusPDUReadCoilsRequest(tag.address, tag.quantity)
        elif isinstance(tag, ModbusTagDiscreteInput):
            pdu = ModbusPDUReadDiscreteInputsRequest(tag.address, tag.quantity)
        elif isinstance(tag, ModbusTagInputRegister):
            pdu = ModbusPDUReadInputRegistersRequest(tag.address, tag.quantity)
        elif isinstance(tag, ModbusTagHoldingRegister):
            pdu = ModbusPDUReadHoldingRegistersRequest(tag.address, tag.quantity)
        else:
            raise NotImplementedError(
                "Modbus tag type not implemented " + str(tag.__class__)
            )

        adu = ModbusTcpADU(
            False,
            self._transaction_generator.increment(),
            self._configuration.unit_identifier,
            pdu,
        )
        write_buffer = WriteBufferByteBased(adu.length_in_bytes(), ByteOrder.BIG_ENDIAN)
        adu.serialize(write_buffer)

        protocol = transport.protocol
        protocol.write_wait_for_response(
            write_buffer.get_bytes(),
            transport,
            adu.transaction_identifier,
            message_future,
        )

        await message_future
        result = message_future.result()

        if isinstance(result, ModbusPDUError):
            response_item = ResponseItem(
                PlcResponseCode.ACCESS_DENIED, PlcNull(result.exception_code)
            )

            response = PlcReadResponse(
                PlcResponseCode.OK, {request.tag_names[0]: response_item}
            )
            return response

        if isinstance(tag, ModbusTagCoil) or isinstance(tag, ModbusTagDiscreteInput):
            a = bitarray()
            a.frombytes(bytearray(result.value))
            a.bytereverse()
            read_buffer = ReadBufferByteBased(bytearray(a), ByteOrder.BIG_ENDIAN)
        else:
            read_buffer = ReadBufferByteBased(
                bytearray(result.value), ByteOrder.BIG_ENDIAN
            )
        returned_value = DataItem.static_parse(
            read_buffer,
            request.tags[request.tag_names[0]].data_type,
            request.tags[request.tag_names[0]].quantity,
        )

        response_item = ResponseItem(PlcResponseCode.OK, returned_value)

        response = PlcReadResponse(
            PlcResponseCode.OK, {request.tag_names[0]: response_item}
        )
        return response

    async def write(
        self, request: PlcWriteRequest, transport: Transport
    ) -> PlcWriteResponse:
        """
        Writes one field from the Modbus Device
        """
        pass
